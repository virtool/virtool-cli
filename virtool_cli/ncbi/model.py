from enum import StrEnum

from pydantic import BaseModel, Field, AliasChoices, field_validator


class NCBIDatabase(StrEnum):
    NUCCORE = "nuccore"
    TAXONOMY = "taxonomy"


class NCBIRank(StrEnum):
    SPECIES = "species"
    ISOLATE = "isolate"


class NCBISource(BaseModel):
    taxid: int = Field(validation_alias="db_xref")
    organism: str
    mol_type: str
    isolate: str = ""
    host: str = ""
    segment: str = ""
    strain: str = ""
    clone: str = ""

    @field_validator("taxid", mode="before")
    @classmethod
    def db_xref_to_taxid(cls, raw: str) -> int:
        return int(raw.split(":")[1])


class NCBINuccore(BaseModel):
    accession: str = Field(validation_alias="GBSeq_primary-accession")
    definition: str = Field(validation_alias="GBSeq_definition")
    sequence: str = Field(validation_alias="GBSeq_sequence")
    source: NCBISource = Field(validation_alias="GBSeq_feature-table")
    comment: str = Field("", validation_alias="GBSeq_comment")

    @field_validator("sequence", mode="after")
    @classmethod
    def to_uppercase(cls, raw: str) -> str:
        return raw.upper()

    @field_validator("source", mode="before")
    @classmethod
    def create_source(cls, raw: list) -> NCBISource:
        for feature in raw:
            if feature["GBFeature_key"] == "source":
                return NCBISource(
                    **{
                        qual["GBQualifier_name"]: qual["GBQualifier_value"]
                        for qual in feature["GBFeature_quals"]
                    }
                )

        raise ValueError("Feature table contains no ``source`` table.")


class NCBILineage(BaseModel):
    id: int = Field(validation_alias=AliasChoices("id", "TaxId"))
    name: str = Field(validation_alias=AliasChoices("name", "ScientificName"))
    rank: str = Field(validation_alias=AliasChoices("rank", "Rank"))


class NCBITaxonomy(BaseModel):
    id: int
    rank: NCBIRank
    lineage: list[NCBILineage]
    species: NCBILineage
