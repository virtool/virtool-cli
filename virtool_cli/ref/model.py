from enum import StrEnum

from pydantic import BaseModel, field_validator


class SubSpeciesBin(StrEnum):
    ISOLATE = "isolate"
    STRAIN = "strain"
    CLONE = "clone"
    VARIANT = "variant"


class IsolateName(BaseModel):
    """Represents a sub-species subcategorization label for sequences"""

    type: SubSpeciesBin
    """The nature of of the subcategory."""

    value: str
    """The name of the subcategory"""

    @field_validator("type", mode="before")
    @classmethod
    def validate_type(cls, raw: str):
        return SubSpeciesBin(raw)
