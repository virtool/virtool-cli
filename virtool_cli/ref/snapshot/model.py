from typing import Annotated, Dict

from pydantic import (
    BaseModel,
    Field,
    TypeAdapter,
    UUID4,
    field_validator,
)

from virtool_cli.utils.models import Molecule
from virtool_cli.ref.utils import IsolateName, IsolateNameType


class OTUSnapshotMeta(BaseModel):
    """Structures metadata about the OTU snapshot itself."""

    at_event: int | None = None
    """The event ID of the last change made to this snapshot."""


class OTUSnapshotSequence(BaseModel):
    """Stores and parses sequence data."""

    id: UUID4
    """The sequence id."""

    accession: str
    """The sequence accession."""

    definition: str
    """The sequence definition."""

    legacy_id: str | None
    """A string based ID carried over from a legacy Virtool reference repository.

    It the sequence was not migrated from a legacy repository, this will be `None`.
    """

    sequence: str
    """The sequence."""

    segment: str
    """The sequence segment."""


class OTUSnapshotIsolate(BaseModel):
    """Stores and parses isolate data."""

    id: UUID4
    """The isolate ID."""

    name: IsolateName
    """The isolate's source name metadata."""

    legacy_id: str | None = None
    """A string based ID carried over from a legacy Virtool reference repository."""

    @field_validator("name", mode="before")
    @classmethod
    def convert_isolate_name(cls, raw: dict) -> IsolateName:
        """Takes a dictionary and converts to IsolateName."""
        return IsolateName(type=IsolateNameType(raw["type"]), value=raw["value"])


class OTUSnapshotOTU(BaseModel):
    """Stores and parses OTU data."""

    id: UUID4
    """The OTU id."""

    taxid: int
    """The NCBI Taxonomy id for this OTU."""

    name: str
    """The name of the OTU (eg. TMV for Tobacco mosaic virus)"""

    acronym: str = ""
    """The OTU acronym (eg. TMV for Tobacco mosaic virus)."""

    legacy_id: str | None
    """A string based ID carried over from a legacy Virtool reference repository."""

    molecule: Annotated[Molecule | None, Field()] = None
    """The molecule of this OTU"""

    segment_schema: Annotated[list | None, Field(alias="schema")] = None

    repr_isolate: UUID4 | None = None


class OTUSnapshotToCIsolate(BaseModel):
    """Stores a table of contents for an isolate."""

    id: UUID4
    """The isolate id."""

    accessions: dict[str, UUID4]
    """A mapping of accessions to sequence ids."""


toc_adapter = TypeAdapter(Dict[str, OTUSnapshotToCIsolate])
