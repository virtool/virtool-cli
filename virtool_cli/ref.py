from pathlib import Path

import click

import virtool_cli.build
import virtool_cli.divide
import virtool_cli.isolate
import virtool_cli.repair
import virtool_cli.taxid


@click.group("ref")
def ref():
    """Commands related to references"""
    pass


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a database reference directory",
)
@click.option(
    "-o",
    "--output",
    default="reference.json",
    help="the output path for a reference.json file",
)
@click.option("-i", "--indent", is_flag=True)
@click.option(
    "-V",
    "--version",
    default=None,
    type=str,
    help="the version string to include in the reference.json file",
)
def build(src_path, output, indent, version):
    """Build a Virtool reference JSON file from a reference directory."""
    try:
        virtool_cli.build.run(Path(src_path), Path(output), indent, version)
    except (FileNotFoundError, NotADirectoryError):
        click.echo("Not a valid reference directory")


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a input reference.json file",
)
@click.option(
    "-o",
    "--output",
    default="src",
    type=str,
    help="the output path for a divided reference directory tree",
)
def divide(src_path, output):
    """Divide a reference.json file from Virtool into a reference directory tree."""
    try:
        if not src_path.endswith(".json"):
            raise TypeError
        virtool_cli.divide.run(Path(src_path), Path(output))
    except (TypeError, FileNotFoundError):
        click.echo(
            "Specified reference file either does not exist or is not a proper JSON file"
        )


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to the input reference.json file",
)
@click.option("-f", "--force_update", is_flag=True)
def taxid(src_path, force_update):
    """Fetch taxid for all OTU in given reference directory."""
    try:
        virtool_cli.taxid.run(Path(src_path), force_update)
    except (FileNotFoundError, NotADirectoryError):
        click.echo("Not a valid reference directory")


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a reference directory",
)
def isolate(src_path):
    """Fetch new isolates for all OTU in a given reference directory."""
    try:
        virtool_cli.isolate.run(Path(src_path))
    except (FileNotFoundError, NotADirectoryError):
        click.echo("Not a valid reference directory")


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a reference directory",
)
def repair(src_path):
    """Fix every OTU in a given reference directory."""
    try:
        virtool_cli.repair.run(Path(src_path))
    except (FileNotFoundError, NotADirectoryError):
        click.echo("Not a valid reference directory")


if __name__ == "__main__":
    ref()
