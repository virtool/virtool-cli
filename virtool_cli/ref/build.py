import json
from pathlib import Path

import arrow

from virtool_cli.ref.legacy import Repo


def build_json(indent: bool, output_path: Path, path: Path, version: str):
    """Build a Virtool reference JSON file from a data directory.

    :param path: The path to a reference repository
    :param output_path: The path to write the output JSON file to
    :param indent: write indented json when enabled
    :param version: the version string to include in the reference.json file
    """
    repo = Repo(path)

    otus = []

    for otu in repo.iter_otus():
        isolates = []

        for isolate in otu.isolates:
            sequences = [
                {
                    "_id": sequence.id,
                    "accession": sequence.accession,
                    "definition": sequence.definition,
                    "host": sequence.host,
                    "segment": sequence.segment,
                    "sequence": sequence.sequence,
                }
                for sequence in isolate.sequences
            ]

            isolates.append(
                {
                    "id": isolate.id,
                    "default": isolate.default,
                    "sequences": sequences,
                    "source_name": isolate.source_name,
                    "source_type": isolate.source_type,
                },
            )

        isolates.sort(key=lambda x: x["id"])

        otus.append(
            {
                "_id": otu.id,
                "abbreviation": otu.abbreviation,
                "isolates": isolates,
                "name": otu.name,
                "schema": otu.schema,
                "taxid": otu.taxid,
            },
        )

    otus.sort(key=lambda x: x["name"])

    with open(output_path, "w") as f:
        json.dump(
            {
                "created_at": arrow.utcnow().isoformat(),
                "data_type": repo.data_type,
                "name": version,
                "organism": repo.organism,
                "otus": otus,
            },
            f,
            indent=4 if indent else None,
            sort_keys=True,
        )
