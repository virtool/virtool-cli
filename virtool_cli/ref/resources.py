import datetime
from dataclasses import dataclass
from pydantic import BaseModel, field_validator
from uuid import UUID

from virtool_cli.ref.utils import DataType, IsolateName, Molecule


class RepoMeta(BaseModel):
    """Represents the metadata for a Virtool reference repository."""

    id: UUID
    """The repository id."""

    created_at: datetime.datetime
    """The date and time the repository was created."""

    data_type: DataType
    """The repository data type."""

    name: str
    """The repository name."""

    organism: str
    """The repository organism."""


@dataclass
class EventSourcedRepoSequence:
    """Represents a sequence in a Virtool reference repository."""

    id: UUID
    """The sequence id."""

    accession: str
    """The sequence accession."""

    definition: str
    """The sequence definition."""

    legacy_id: str | None
    """A string based ID carried over from a legacy Virtool reference repository.

    It the sequence was not migrated from a legacy repository, this will be `None`.
    """

    sequence: str
    """The sequence."""

    segment: str
    """The sequence segment."""

    def dict(self):
        return {
            "id": self.id,
            "accession": self.accession,
            "definition": self.definition,
            "legacy_id": self.legacy_id,
            "sequence": self.sequence,
            "segment": self.segment,
        }


@dataclass
class EventSourcedRepoIsolate:
    """Represents an isolate in a Virtool reference repository."""

    id: UUID
    """The isolate id."""

    legacy_id: str | None
    """A string based ID carried over from a legacy Virtool reference repository.

    It the isolate was not migrated from a legacy repository, this will be `None`.
    """

    name: IsolateName
    """The isolate's source name metadata."""

    sequences: list[EventSourcedRepoSequence]
    """A list of child sequences."""

    @property
    def accession_set(self) -> set:
        """Return a set of accessions contained in this isolate"""
        return {sequence.accession for sequence in self.sequences}

    def dict(self):
        return {
            "id": self.id,
            "legacy_id": self.legacy_id,
            "name": self.name.model_dump(),
            "sequences": [sequence.dict() for sequence in self.sequences],
        }


@dataclass
class EventSourcedRepoOTU:
    """Represents an OTU in a Virtool reference repository."""

    id: UUID
    """The OTU id."""

    acronym: str
    """The OTU acronym (eg. TMV for Tobacco mosaic virus)."""

    excluded_accessions: set[str]

    isolates: list[EventSourcedRepoIsolate]
    """A list of child isolates."""

    legacy_id: str | None
    """A string based ID carried over from a legacy Virtool reference repository.

    It the OTU was not migrated from a legacy repository, this will be `None`.
    """

    name: str
    """The OTU name (eg. Tobacco mosaic virus)."""

    taxid: int
    """The OTU taxonomy ID."""

    molecule: Molecule | None = None
    """The molecule of this OTU"""

    schema: list | None = None
    """The schema of the OTU"""

    @property
    def accession_set(self) -> set:
        """Return a set of accessions contained in this isolate"""
        accessions = set()
        for isolate in self.isolates:
            accessions.update(isolate.accession_set)

        return accessions

    @property
    def blocked_accession_set(self) -> set:
        """Returns a set of accessions to be blocked from fetches,
        i.e. accessions that have already been added and excluded accessions."""
        return self.accession_set.union(self.excluded_accessions)

    def get_isolate(self, isolate_id: UUID) -> EventSourcedRepoIsolate | None:
        """Return the isolate instance associated with a given UUID if it exists,
        else None.
        """
        for isolate in self.isolates:
            if isolate.id == isolate_id:
                return isolate

        return None

    def get_isolate_id_by_name(self, name: IsolateName) -> UUID | None:
        """Return an UUID if the name is extant in this OTU."""
        for isolate in self.isolates:
            if isolate.name == name:
                return isolate.id

        return None

    def dict(self):
        return {
            "id": self.id,
            "acronym": self.acronym,
            "excluded_accessions": self.excluded_accessions,
            "isolates": [isolate.dict() for isolate in self.isolates],
            "legacy_id": self.legacy_id,
            "name": self.name,
            "molecule": self.molecule,
            "schema": self.schema,
            "taxid": self.taxid,
        }
