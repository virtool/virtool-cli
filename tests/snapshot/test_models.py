import shutil

import pytest
from syrupy import SnapshotAssertion

from virtool_cli.ref.snapshot.model import (
    OTUSnapshotOTU,
    OTUSnapshotIsolate,
    OTUSnapshotSequence,
)

from virtool_cli.ref.resources import (
    EventSourcedRepoOTU,
    EventSourcedRepoIsolate,
    EventSourcedRepoSequence,
)


@pytest.fixture()
def snapshotted_scratch(scratch_repo):
    scratch_repo.snapshot()

    yield scratch_repo

    shutil.rmtree(scratch_repo.cache_path / "snapshot")


class TestRepoToSnapshotModel:
    @pytest.mark.parametrize(
        "taxid,accessions",
        [
            (
                438782,
                [
                    "NC_010314",
                    "NC_010315",
                    "NC_010316",
                    "NC_010317",
                    "NC_010318",
                    "NC_010319",
                ],
            ),
        ],
    )
    def test_sequence_conversion(
        self, taxid, accessions, snapshotted_scratch, snapshot: SnapshotAssertion
    ):
        otu = snapshotted_scratch.get_otu_by_taxid(taxid)

        for accession in accessions:
            original_sequence = otu.get_sequence_by_accession(accession)

            assert type(original_sequence) is EventSourcedRepoSequence

            converted_model = OTUSnapshotSequence(**original_sequence.dict())

            assert converted_model.model_dump() == snapshot

    @pytest.mark.parametrize("taxid", [438782, 1441799, 430059])
    def test_isolate_conversion(
        self, taxid, snapshotted_scratch, snapshot: SnapshotAssertion
    ):
        otu = snapshotted_scratch.get_otu_by_taxid(taxid)

        for isolate in otu.isolates:
            assert type(isolate) is EventSourcedRepoIsolate

            converted_model = OTUSnapshotIsolate(**isolate.dict())

            assert converted_model.model_dump() == snapshot

    @pytest.mark.parametrize("taxid", [438782, 1441799, 430059])
    def test_otu_conversion(
        self, taxid, snapshotted_scratch, snapshot: SnapshotAssertion
    ):
        otu = snapshotted_scratch.get_otu_by_taxid(taxid)

        assert type(otu) is EventSourcedRepoOTU

        converted_model = OTUSnapshotOTU(**otu.dict())

        assert converted_model.model_dump(by_alias=True) == snapshot
