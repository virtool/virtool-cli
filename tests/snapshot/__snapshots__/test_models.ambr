# serializer version: 1
# name: TestRepoToSnapshotModel.test_isolate_conversion[1441799]
  dict({
    'id': UUID('08bcc1e3-d606-4b36-992b-7fcfeb588010'),
    'legacy_id': None,
    'name': dict({
      'type': isolate,
      'value': 'Peshtatuek_12b',
    }),
  })
# ---
# name: TestRepoToSnapshotModel.test_isolate_conversion[430059]
  dict({
    'id': UUID('b1505673-c45e-4c83-93a0-7f5bceb67b71'),
    'legacy_id': None,
    'name': dict({
      'type': refseq,
      'value': 'NC_043170',
    }),
  })
# ---
# name: TestRepoToSnapshotModel.test_isolate_conversion[438782]
  dict({
    'id': UUID('98371715-6d50-40a1-a77e-a168105eddbd'),
    'legacy_id': None,
    'name': dict({
      'type': isolate,
      'value': 'Q767',
    }),
  })
# ---
# name: TestRepoToSnapshotModel.test_isolate_conversion[438782].1
  dict({
    'id': UUID('af7602c9-d636-4185-a2f3-ebd1f518979c'),
    'legacy_id': None,
    'name': dict({
      'type': isolate,
      'value': 'Q1108',
    }),
  })
# ---
# name: TestRepoToSnapshotModel.test_otu_conversion[1441799]
  dict({
    'acronym': '',
    'id': UUID('c2a22914-12cf-4b66-aa42-e1395e4498de'),
    'legacy_id': None,
    'molecule': None,
    'name': 'Faba bean necrotic stunt alphasatellite 1',
    'repr_isolate': None,
    'schema': list([
    ]),
    'taxid': 1441799,
  })
# ---
# name: TestRepoToSnapshotModel.test_otu_conversion[430059]
  dict({
    'acronym': '',
    'id': UUID('352325fa-bff2-44b1-a5ad-64de5c786c74'),
    'legacy_id': None,
    'molecule': None,
    'name': 'Vallota mosaic virus',
    'repr_isolate': None,
    'schema': list([
    ]),
    'taxid': 430059,
  })
# ---
# name: TestRepoToSnapshotModel.test_otu_conversion[438782]
  dict({
    'acronym': '',
    'id': UUID('1f88507e-1a9a-4ec7-88ed-cc253c6200ce'),
    'legacy_id': None,
    'molecule': None,
    'name': 'Abaca bunchy top virus',
    'repr_isolate': None,
    'schema': list([
    ]),
    'taxid': 438782,
  })
# ---
# name: TestRepoToSnapshotModel.test_sequence_conversion[438782-accessions0]
  dict({
    'accession': 'NC_010314',
    'definition': 'Abaca bunchy top virus DNA-N, complete genome',
    'id': UUID('1c44278c-8283-438a-83af-090b4bbc4056'),
    'legacy_id': None,
    'segment': 'DNA N',
    'sequence': 'AGCAGGGGGGCTTATTATTACCCCCCCTGCTCGGGGCGGGACATTCTGTGATGGGCTGGGCTTTATGCGGCCAAATAAGCCCATAAAGCCAGATCTGGGCCCATTTAAGGGCCCGTGGTTTGAAAATGTCGCGTTCCCGCCTAAATTGTTTGCTTGCCCTGCAAGGAAACGAAAACTCTATAAATAGGGTTGTTCTCTGCTTGTTTAATACATCAGGCGCAAATCTTTTGCAACGATGGATTGGATGGAATCACAATTCAAGACATGTACGCATGGCTGCGACTGGAAGGCGATAGCTCCAGAAGCACAAGATAATATACAGGTAATTACATGTTCCGATTCAGGTTACGGAAGAAAGAACCCTCGTAAGGTTCTTCTGAGGAGTATTCAGATAGGGTTCAATGGAAGCTTCAGAGGAAGTAATAGAAATGTTCGAGGCTTCATATACGTGTCTGTAAGACAGGATGATGGCCAAATGAGACCAATTATGGTCGTTCCATTCGGAGGGTATGGATATCATAACGACTACTATTATTTTGAAGGACAATCCAGTACGAATTGTGAGATAGTGTCGGACTATATTCCGGCCGGTCAAGACTGGAGCAGAGATATGGAGATAAGTATAAGTAACAGCAACAATTGTAATCAAGAGTGCGATATCAAGTGTTATGTAGTATGTAATTTAAGAATTAAGGAATAAWATTGTTGCCGAAGGTCTGTTATTTGAATGTTGAGATAAGGAAAGGGGCGGCGAAGCATGTGTGTATAATAACATATAACACACTATTATATATTTTGTAAAGAATAAAATTATGACCTGTCAGATTAAGTTTAGAATGAACTGAGGCCGAAGGCCTCACCGAGGCCGAAGGCCGTCAGGATGGTTTTACAAAATAATTATAAGCACCTGTACTAAGTACGAAGAGCGGTATAATATCTGAAAGGAAAAAATAATAATATAATAAAAATATTATGATGTCCCAAAATAGCAGAATGCTAAAGGAACAAAAGGATGCTCTAAGTACAGGGTTGCGTGCTCTGGACGCCACTTTAGTGGTGGGCCAGATGTCCCGAGTTAGTGCGCCACGTC',
  })
# ---
# name: TestRepoToSnapshotModel.test_sequence_conversion[438782-accessions0].1
  dict({
    'accession': 'NC_010315',
    'definition': 'Abaca bunchy top virus segment 2, complete sequence',
    'id': UUID('692475d4-056a-4680-aa74-b772816c314c'),
    'legacy_id': None,
    'segment': 'DNA U3',
    'sequence': 'AGCAGGGGGGCTTATTATTACCCCCCCTGCTGGGACGGGACATCCGAATAGTAATGGGCTTAATACATAAATGGGCCGCCAAATAAGCGTACAGTTCAGTATCTTCGTTTTGGGCCTCGGCCCAAAATTAAGAGAACGTGTGCGCTTGTTTGGGTTGGACCGTAGGTCCGGTTCGAATGAGGAATATTCCTTGCTTCCATAGGAAGGAACAGCGAAATAAAATAAATCGCTGTTACACGTGTAAGAATACTATTAGTTCGCAGAGTTTATAAATACCCTTCAAGTTGAAGGGTGTGGTGCTCTCTCTCTTCTGTCAGACGGTGTGCCATGATGAAGCTCTCCGGGAGGAAAGAAGGGCGGAGGTGGAGGAGCCGCCTCGCCCGTCCCTCCGACGAAGTATACGAAGTCGTAGGTATGTCAGGGTATTTATAGAAGCGTCGTATACTAAACGCGCCTGTATCATAATTGTATTTGTGTATTTAAATATTTAAATACCAAACCTTAAGGAATAAAATATATGTTTAATAAAATAACACATGACAATAAAAGAAAACAATAATACAATAATAAAATTGTATTATGCAAACGTTGTATAATTAAGGTCTGTTTAATATACATATATATATGTATATTACAGTATTGTTTAAAATAAATGACTTGGAAGGAAATAATAATATTAATGATAAGCAATAAAATATTCCGAATACATAAAAAGGGGAAAATGCAATTAAGACCTGTCACGTTTACATATTTCGAATATATTAAGGGCCGAAGGCCCGTCAGTATGCAGGTACATCAGTGATTGCTTCGTGACGAAGCAAGGGTATTATTGTAAATAAGAAAGCATCTGACAACTTTAATAGTGGTCCCCTATACAGCTGTCATATGACAGCTGGCAAAGGATCATTGGGCGGACTCCAAATATATATTAAATATAACATATAAAATATATAAGGTATAGATACTATTATTAATTAAAAGAGGCGGGAAAGAGGACACGTGCGCCGCACGCCACTTTAGTGGTGGGCCATATGTCCCGAGTTAGTGCGCTTACGTC',
  })
# ---
# name: TestRepoToSnapshotModel.test_sequence_conversion[438782-accessions0].2
  dict({
    'accession': 'NC_010316',
    'definition': 'Abaca bunchy top virus DNA-S, complete genome',
    'id': UUID('7462ad59-81a4-4ac1-b00e-e2ef6e32ead6'),
    'legacy_id': None,
    'segment': 'DNA S',
    'sequence': 'GGCAGGGGGGCTTATTATTACCCCCCCTGCCCGGGACGGGACATTCTGTGATGGGCTGGGCTTTATGCGGCCAAATAAGCCCATAAAGCCAGATCTGGGCCCATTTAAGGGCCCGTGGTTTGAAAATGTCGCGTTCCCGCCAAAATAGTTGCTTGCTCTGCAAGCAAACTATATCTATTATAAATACCAGATGTAATGGTTGCCTGAGAATGAAAATAAAAGGATGGCGAGGTATCCGAAGAAGGCGCTAAAGAAGAGGAAGGCGGTACGCCGTAAGTATGGAAGTAAAGCTACGACCAGTCATGATTATGCTGTTGATACTTCATTTATTGTTCCTGAAAATACTATTAAGCTGTATCGTATTGAGCCTACTGATAAAACATTACCCAGATATTTTATCTGGAAAATGTTTATGTTGTTGGTATGCAAAGTCAGACCAGGGCGTATACTTCATTGGGCGATGATTAAGAGCTCTTGGGATGTGAAGGATCCAACAGTTGTTCTTGAAGCGCCTGGATTGTTTATCAAGCCAGCAAACAGCCATCTGGTTAAACTGGTATGCAGTGGCGAGTTAGAAGCTCCAGTAGGAGGAGGGACTTCAGAGGTTGAATGTCTTCTACGGAAGACAACTTTACTTCGTAGGAATGTTACAGAATTGGATTTCTTGTATTTGGCGTTTTATTGTTCTTCTGGAGTTACAATCAACTACCAGAACAGGATTACATATCATGTATAAACCACATAAAATAAATGTGGTGTTGCAGGCATGGGAAGAATAAAACAATGTTTGCCTACGAAATATTTTGGTAAAGTGAAATTATGACCTGTCAGAATTAAGTTTAGAATGAACTGAGGCCGAAGGCCTCACCGAGGCCGAAGGCCGTCAAGTTGGATGAATAAAATACAAGGTATAAGTACGAAGAGCGGTATAATATCTGAAAGGAAATAATAATAATATAATAAAATATTATGATGTCCCAAAATAGCAGAATGCTAAAGGAACAAAAGGATGCTCTAAGTACAGGGTTGCGTGCTCTGGACGCCACTTTAGTGGTGGGCCAGATGTCCCGAGTTAGTGCGCCACGTG',
  })
# ---
# name: TestRepoToSnapshotModel.test_sequence_conversion[438782-accessions0].3
  dict({
    'accession': 'NC_010317',
    'definition': 'Abaca bunchy top virus DNA-M, complete genome',
    'id': UUID('afc663a9-49b1-4c54-8435-27e59caa24f0'),
    'legacy_id': None,
    'segment': 'DNA M',
    'sequence': 'GGGGCTGGGGCTTATTATTACCCCCAGCCCCGGAACGGGACATCACGTGCCTCCCTGTACACGCACGTGACATCAAATGGCTTGCTGYGCAAGATAGAAAGGCTTCCTGCGGAAGCCAGGCATCATATATGGGTTGTAGAGGATCAGCCGACTCTATAAATATAGGGAGGCGTGGTCATGGGATTGACAGGAGAACGAGTGAAACAATTCTTTGAATGGTTTCTGTTCTTCTCAGCAATATTTGTTGCAATAACAATAATATATATATTGCTGGCAGTTCTCCTTGAACTACCGAAGTATATTAAGGGTTTAGTACAATACGTAGTGGAGTATATTACTAGACGACGGGTATGGACACGGAGGACCCAATTGACAGAGGCAACCGGAGGAGGCGATATAGAAGCTGTAGGGCATGACAGTCAGGCGTATACGCATACTGTTATGCCATSTGTTCCACCAGTTAGCGCTCCTATATCAAATAGGAGAGCTGATCAGCCTCTTCGATCAAGCGCCGGACCAATGTTTTAAATACCCGTGATATTTAATATGCAAGTGTATAAATACCCATTGTAGATCTGTTTGTAACCTGAATATGCAAAGTATATAATACTTTGTTGTAATGTATAAGTACATTAATAATATACGAAGTATAATGTTGATGCGATGTCTTCGGAAAATGAAGTATACCCAAATACACAAAAAAACATATATGTGGTGTATACTTATTGTTAAGTATAATAAAATTATAATACAAACAAATATATTGTGTATTATAATACATAAAAGAAGACAGAGCTGTGAAGTTAAGTAAGAAAGCGACGGATTCGTATTGGATAATATGATTCGCGGAGCATTACTTAACGGCGAAGTAAAGCATCAGACAAGAATATGACAGCTGTCATATCAACTAAAAAGCATAGCTTGCCGAGCACGCTATGCAATATAAGGGAAAAATGAAATAATGACCTGTCACGTTTACACTTTTCGTAAAGATGTAGGGCCGAAGGCCCTAATGACGCGTGTCATATTCTCTATAGTGGTGGGTCATATGTCCCGAGTTAGTGCGCCACGTAA',
  })
# ---
# name: TestRepoToSnapshotModel.test_sequence_conversion[438782-accessions0].4
  dict({
    'accession': 'NC_010318',
    'definition': 'Abaca bunchy top virus DNA-C, complete sequence',
    'id': UUID('bc0c9c5d-8e7d-4440-bec0-3fcb5584f792'),
    'legacy_id': None,
    'segment': 'DNA C',
    'sequence': 'GGCAGGGGGGCTTATTATTACCCCCCCTGCCCGGGACGGGACATCACGTGCGCCTATAAAGAAGCACGTGATGATGTGACATATGTTTGCCGTAAGCTGTACAAAAGCGAATGCTTTATTGCTTTATTCGTTGCTTAGCCCCTAAGCAACCTGGTAGTGGGATGTGGTCCCTGATGCGAAGCAGCTTTGTTGTTGCTATTTATATCTGTGTTCGCCATGTAAAATGCGAAATACAACGCTGATCAGAATGGAGTTCTGGAATTCGGAAGCATTTTGCGACGATGTGAAGCGTGTCATTAAGCAAAAATATTGGGAAGAGCGGATGAAATCTCTATTTATAGAGAAGGTGAGTGGTTATGTTCGAAGGATTCTTGTTTATGGTAATCTTGATGATACCATATATGCGGTTCAGCAAATGAAGACTTCTATAGTTCAGTGTGCTGAACGTTTCGGTAAAGCCTGTGTGGTGGTATATAATGGTTTGGATCCATCAATAGGTTTCCGATTACACACTATGGCGTTCTTCTTCGAAGAATATGTTGAGGAAGTGAGTACTGCAGATCCAATGGCGGTTCAGTTATTTTGTGATGAAGAAATAGAAGAATTCTCAAATTCTGATGTACGCCTTATTAAAAATGTTATTATGGCATCGACAGATGCATCGATTGATGTAGGAAATTGTATTCAGATAATAATATCTGATAATGTAATAACATTCTATATATGTTAACTTCATTTATACATAAGAATGAATGAAGTGGTTTATTTATGATTTTTAGAATATAATCATAAATGGTAAACCTTAAGCAAAACCACCCTAAAACAAATAAACCTCTCTGACATACAACCCTCTATAAAATAAAGCCCATGTAAGATTCAAATTTAGAATGAAAAATGGGCCGAAGGCCCATATAAATGCATTTAAGGCCGAAGGCCTTATAATTGCAGGAAGAAAAGAACACGGTTTTGCTTACGTGGCCTGTGGGCCATATGTCCCGAGTTAGTGCGCCACGTG',
  })
# ---
# name: TestRepoToSnapshotModel.test_sequence_conversion[438782-accessions0].5
  dict({
    'accession': 'NC_010319',
    'definition': 'Abaca bunchy top virus DNA-R, complete genome',
    'id': UUID('d2da7973-9621-4b6c-8e0d-97580a79d9e2'),
    'legacy_id': None,
    'segment': 'DNA R',
    'sequence': 'GGCAGGGGGGCTTATTATTACCCCCCCTGCCCGGGACGGGACATTTGCATCTATAAATAGAAGCGCCCTCGCTCAACCAGATCAGGCGCTGCAATGGCTAGATATGTCGTATGTTGGATGTTCACCATCAACAATCCCGAAGCTCTTCCAGAGATGAGGGAAGAATACAAATACCTGGTTTACCAGGTGGAGCGAGGCGAAAGCGGTACACGACATGTGCAGGGCTATGTTGAAATGAAGAGACGAAGTTCTCTGAAACAAATGAGGGCTTTAATTCCTGGTGCCCATCTCGAAAAGAGAAGGGGCACACAGGAAGAAGCTAGAGCTTATTGTATGAAGGCAGATACGAGAGTCGAAGGTCCCTTCGAGTTTGGTCTTTTCAAAGTATCATGTAATGATAATTTGTTTGATGTCATACAGGATATGAGAGAAACGCACAAACGGCCGATTGAGTATTTATACGACTGTCCTAATACCTTCGATAGAAGTAAGGATACATTATACAGGGTACAAGCGGAAATGAATAAAATGCAAGCTATGATGTCGTGGTCGGAAACCTATGGTTGCTGGACGAAGGAAGTGGAGGAACTAATGGCGGAGCCATGTCACCGACGGATTATTTGGGTCTATGGCCCAAATGGTGGTGAAGGTAAAACAACCTATGCGAAGCATCTAATCAAGACCAGAAATGCATTTTATACACCTGGCGGAAAGACACTGGATATATGTAGGCTGTATAATTATGAGGGAATTGTAATATTTGATATTCCCAGATGCAAAGAGGATTACTTGAATTACGGAATTCTTGAGGAATTCAAGAATGGCATCATTCAGAGCGGGAAATATGAACCAGTTTTAAAAATTGTAGAGTATGTGGAGGTCATTGTCATGGCTAACTTCCTGCCGAAGGAAGGAATATTCTCGGAAGACCGAATAAAGCTTGTAACTTGTTGAACACGCTATGCAATAAAGGGGAAAAATGCAATTATGACCTGTCACGTTTACACTTTTCGTAAAGATGTAGGGCCGAAGGCCCTAATGACGCGTGTCATATTCTCTATAGTGGTGGGTCATATGTCCCGAGTTAGTGCGCCACGTG',
  })
# ---
